js/source/tile_pyramid.js

```diff
@@ -16,24 +17,65 @@ module.exports = TilePyramid;
  * @private
  */
-function TilePyramid(options) {
-    this.tileSize = options.tileSize;
-    this.minzoom = options.minzoom;
-    this.maxzoom = options.maxzoom;
-    this.roundZoom = options.roundZoom;
-    this.reparseOverscaled = options.reparseOverscaled;
-
-    this._load = options.load;
-    this._abort = options.abort;
-    this._unload = options.unload;
-    this._add = options.add;
-    this._remove = options.remove;
-    this._redoPlacement = options.redoPlacement;
+function TilePyramid(id, options, dispatcher) {
+    this.id = id;
+    this.dispatcher = dispatcher;
+
+    var onChange = function () {
+        this.reload();
+        if (this.transform) {
+            this.update(this.transform, this.map && this.map.style.rasterFadeDuration);
+        }
+    }.bind(this);
+
+    // TODO: this is an ugly consequence of inverting tilepyramid / source
+    // dependency and making source creation async: we use a dumb, generic
+    // version of serialize() until the source is loaded at which point we
+    // switch to deferring to the source's method.
+    // Options:
+    // 1. Make Source.create synchronous.  Drawback: would require adding `loaded()` or `onLoad`  to the Source interface.
+    // 2. Remove Source:serialize() entirely, and instead let TilePyramid:serialize() always jut return the options object it was created with.
+    // 3. Leave the below wart as-is.
+    // I prefer option 2 if it's workable.
+    this.serialize = function () { return options; };
+
+    var sourceLoaded = function (err, source) {
+        if (err) {
+            this.fire('error', {error: err});
+            return;
+        }
+
+        // TODO: remove this once we eliminate need for sources to have a
+        // reference to the Map instance.
+        this._source = source;
+        if (this.map && this._source.onAdd) { this._source.onAdd(this.map); }
+
+        this._sourceLoaded = true;
+
+        this.tileSize = source.tileSize;
+        this.minzoom = source.minzoom;
+        this.maxzoom = source.maxzoom;
+        // TODO: not exactly sure on these three properties and whether they
+        // make sense as part of the Source interface.
+        this.roundZoom = source.roundZoom;
+        this.reparseOverscaled = source.reparseOverscaled;
+        this.isTileClipped = source.isTileClipped;
+
+        // TODO: what's cleaner? 1) grab references to the Source methods we
+        // want, or 2) just keep a reference to the Source itself and delegate?
+        var noop = function () {};
+        this._load = (source.load || noop).bind(source);
+        this._abort = (source.abort || noop).bind(source);
+        this._unload = (source.unload || noop).bind(source);
+        this.serialize = (source.serialize || noop).bind(source);
+        this.prepare = source.prepare && source.prepare.bind(source);
+
+        this.fire('load');
+    }.bind(this);
+
+    Source.create(id, options, dispatcher, onChange, sourceLoaded);
```

Comment:
It seems like there are a number of chicken-and-egg problems with `Source`s and `TilePyramids` right now. What do you think of this design:

`TilePyramid` (maybe renamed to `SourceCache`) is treated as a [wrapper / decorator](https://en.wikipedia.org/wiki/Decorator_pattern) to `Source`s


```js
var source = new GeoJSONSource(...);
source.loadTile(...); // does all kinds of HTTP fetching and tile generation
var sourceCache = new SourceCache(...);
sourceCache.loadTile(...); // immediately returns tile from the cache if possible, otherwise sends the request to the underlying source
```

I admit that the current `TilePyramid` doesn't perfectly map onto the proposed `SourceCache` class but I think we can make it work / ticket out some tail work.